### PR TITLE
Dev command in non interactive mode

### PR DESCRIPTION
**Pull request by alvaro-shopify**
### WHY are these changes introduced?
As a requirement for running the `dev` command inside a `spin instance` in an automatic way, some modifications should be added to support this feature.

### WHAT is this pull request doing?
- Add new flags to the `dev` command:
  - `--api-secret`: The API secret of your app
  - `--include`: Only the extension which local identifier matches this value will be included in the dev flow execution
  - `--exclude`: All the existing extensions will be included in the dev flow execution except for the one which local identifier matches this value
  - Fetch remote extension specifications skipped in case the command is run in `spin` as a service
  - Add lazy fetching of the extension specifications. Specs are only loaded in case the app includes at least one extension
  - Modified every command that `load` the app to use lazy fetching of the extension specifications so that operation is skipped in case the app has no extensions.

`--include` and `--exclude` flags cannot be used simultaneously. In that case the cli will return an error

image

Extensions identifiers used with `--include` and `--exclude` flags should exists. An error is returned in case at least one does not exist:

image

### How to test your changes?
- Write down the `Client ID` and `Client secret` from one of your app using the `Partners dashboard` 

- Update app url manually with the ngrok tunnel you are going to use. In spin 1P party apps have fixed app urls so there is no needing on update them that involves prompting the user

- Right now it's not possible to run `dev` command as a service (non-interactive mode) in `spin` for app with extensions, so all of them should be removed from the `dev` workflow using the new `--exclude` flag. 
This is an example of how to run the `dev` command in a non-interactive mode with two extensions 

```
SPIN=1 SHOPIFY_FLAG_APP_API_KEY=8289684b401bbac67177ec6728d120b3 SHOPIFY_FLAG_APP_API_SECRET=c6fd90b7d741d8c331eb0b1ce355b615 SHOPIFY_FLAG_STORE=cli-second.myshopify.com pnpm shopify app dev --path ~/cli-apps/last-node2 --no-update --tunnel-url https://9b50-47-62-197-187.ngrok.io:5675 --exclude checkout-02 --exclude theme-app-extension 
```

### Post-release steps

- Modify the documentation to add the new parameter
- Add new instrumentation information about the usage of the `non-interactive dev mode`

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [ ] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

**Comment by pepicrft** [comment]
Hey @alvaro-shopify
Introduce a boolean that we pass around to take different paths executions depending on its value will make the code harder to reason about and test. For instance, what's the connection between "interactivity" and "updating or not URLs"? Taking the use case that you included in the description of the PR:

> Skip dev for apps with theme app extension as right now it's not possible to be run without prompting for identification

What about having a more generic flag for projects to skip the dev'ing of some of the blocks?

```
yarn shopify app dev --exclude my-theme-app-extension --exclude other-extension
```
cc @MeredithCastile to help with the usage and naming of flags. Should we have also the `--only` version? What happens if we use the two?

> In this mode, urls should be updated manually. In case of the 1P apps in spin, they are already configured with the correct fixed url

Same here, if "should be updated manually" is something that the user needs to control, then we should have a flag for that. Otherwise the "In this mode, urls should be updated manually. In case of the 1P apps in spin, they are already configured with the correct fixed url" will be implicitly codified in code that will be hard to maintain.

**Review by pepicrft — CHANGES_REQUESTED** [review]
I'd revisit the approach with @MeredithCastile and use flags to control the behaviour of the CLI.

**Comment by alvaro-shopify** [comment]
> That's what I'd suggest @alvaro-shopify. I'd recommend bringing this up in a conversation with @MeredithCastile. Regarding the last point, I'll check how other CLIs are supporting "disabling interactivity". Out of curiosity, does the CLI detect the environment as interactive when it's being launched by Spin?

Right now in this PR, regardless the environment where the CLI is run, the flow is "flagged" as non-interactive in case api-key, api-secret and store are provided using the corresponding flags/env-variables.

Please @MeredithCastile, could you take a look at this when possible?. We need to cover this requirement to finish the migration of the  1P app `forms` project. I think they will test this behaviour next week using a patched CLI version, so we have some time

**Comment by alvaro-shopify** [comment]
@pepicrft @MeredithCastile 
I have made the changes proposed by @pepicrft . Instead of adding a specific logic to interpret what is an `interactive mode` and what steps should be skipped in that `interactive mode`, I have added some flags to the `dev` command that allows the user to control what should be skipped and what not.
- `--skip-remote-specifications`: Use only local extension specification instead of fetching the remote ones that needs a `partner token` (needs the user to identify)
- `--skip-extension-category`: Allows the user to skip `deving` extension types categories. It's a multiple value flag that admits 'ui', 'theme' and 'function'

So right now to run the `dev` command in a non-interactive way the command should look like this:
```
SHOPIFY_FLAG_APP_API_KEY=#your_client_id# SHOPIFY_FLAG_APP_API_SECRET=#you_client_secret# SHOPIFY_FLAG_STORE=#your_store_name# app dev --skip-remote-specifications --no-update --skip-extension-category ui --skip-extension-category theme
```

When `profiles` are ready we can add a default one for `spin` environments that set the flags 
```
  --skip-remote-specifications 
  --no-update 
  --skip-extension-category ui 
  --skip-extension-category theme
```

**Comment by MeredithCastile** [comment]
Thanks for looping me in. The longer-term solution may warrant a synchronous conversation, but I'll do my best here.


### Overall short-term approach

> SHOPIFY_FLAG_APP_API_KEY=#your_client_id# SHOPIFY_FLAG_APP_API_SECRET=#you_client_secret# SHOPIFY_FLAG_STORE=#your_store_name# app dev --skip-remote-specifications --no-update --skip-extension-category ui --skip-extension-category theme

I understand why we've arrived at the above solution, but there's tremendous potential for user error here (copying and pasting so many values). Shopify developers deserve a better experience than this.

I gather that this is a bandaid solution. This is the first time I'm hearing of "Profiles"; is that related to Presets?


### Potential improvements to the overall short-term approach

- Remove "FLAG" throughout
- Remove "APP" throughout (?)
- Remove "API" before "Secret" (for whatever reason, our established terminology is "API key" and just "secret")
- Use "extension" rather than "specification" (ex: skip-remote-extensions) — rationale is that extensions are the primary conceptual category; would that work?
- Make the last flag specific to theme app extensions since that's the only use-case here (--skip-theme-app-extensions)
- Potentially create an alias to *add* URL to the update flag to make its purpose more obvious in the logs

So like this:

> SHOPIFY_API_KEY=#your_client_id# SHOPIFY_SECRET=#your_client_secret# SHOPIFY_STORE=#your_store_name# app dev --skip-remote-extensions --no-URL-update --skip-theme-app-extension

___

Do we have a strategy for adding this to the Dev Handbook and other scaled instructions?

**Comment by pepicrft** [comment]
I agree with @MeredithCastile that this should be better discussed in a meeting. Otherwise, might might end up doing work that we'll need to revert. @alvaro-shopify do you mind scheduling a quick call with the two of us to discuss this further?

**Comment by MeredithCastile** [comment]
Notes from a call with Pedro and Alvaro just now:

### Decision 1.
For the Spin use-case, we'll eliminate the step around local/remote matching. We can handle that behind the scenes. That decision is make.

### Decision 2.
For the case in which a developer just wants to look at a piece of their app, we'll enable that filtering via flags/categories/IDs.

Still to do: determine precisely how that filtering will work.

Pedro raised a few examples:
- [pnpm wildcard filtering by category](https://pnpm.io/filtering) (pnpm --filter "@babel/*" test)
- Cargo lets you scope down to only one package (cargo build --package foo   =  # build the package foo)
- yarn is simply --include and --exclude

Here's what the wildcard approach might look like in our system: `yarn shopify app dev --filter "function-*"`. I see why the wildcard functionality is appealing, but I'm inclined to save that for a later version once we have the categories in our system more codified. I'm not sure everyone on the Product side is on board with "theme app extensions" being conceptually distinct from "ui extensions." 

**My recommendation, then would be 2 flags: --include=EXTENSION_ID, --exclude=EXTENSION_ID.** For simplicity, given the fact that app IDs aren't on the same system as extension IDs, we can just always dev the app home (web). We can throw an error if the command has an include and exclude flag with the same ID.

Error if the ID isn't found: "Extension ID not found. Make sure you have the right extension ID, then run 'dev' again." <-- @pepicrft, any suggestions on making this more helpful. Where do 1P devs find their extension IDs?"

**Comment by pepicrft** [comment]
> My recommendation, then would be 2 flags: --include=EXTENSION_ID, --exclude=EXTENSION_ID. For simplicity, given the fact that app IDs aren't on the same system as extension IDs, we can just always dev the app home (web). We can throw an error if the command has an include and exclude flag with the same ID.


@MeredithCastile by `EXTENSION_ID` are you referring to the unique long IDs that we generate on Partners? Or are you referring to the local ID which is the name of the folder containing the extension?

**Comment by Arkham** [comment]
just a random thought: `--exclude` makes sense to me, but I prefer `--only` to `--include`

**Comment by alvaro-shopify** [comment]
> just a random thought: `--exclude` makes sense to me, but I prefer `--only` to `--include`

But does it make sense to use both flags at the same time?

**Comment by Arkham** [comment]
no, we should check that and allow one of them at a time

File: packages/app/src/cli/services/dev.ts [hunk 1]
```diff
@@ -397,4 +412,64 @@ async function logMetadataForDev(options: {
...
+        )
+    })
+
+  validateExistingExtensions([...extensionsIncluded, ...extensionsExcluded])
+
+  const included = (extensions: Extension[], filter: string[]) =>
+    extensions.filter((extension) => filter.includes(extension.localIdentifier))
+  if (extensionsIncluded && extensionsIncluded.length > 0) {
+    app.extensions.function = included(app.extensions.function, extensionsIncluded) as FunctionExtension[]
+    app.extensions.ui = included(app.extensions.function, extensionsIncluded) as UIExtension[]
```
Comment (isaacroldan) [inline packages/app/src/cli/services/dev.ts:458]:
```suggestion
    app.extensions.ui = included(app.extensions.ui, extensionsIncluded) as UIExtension[]
```

File: packages/app/src/cli/services/dev.ts [hunk 2]
```diff
@@ -397,4 +412,64 @@ async function logMetadataForDev(options: {
...
+    app.extensions.theme = included(app.extensions.theme, extensionsIncluded) as ThemeExtension[]
+    return app
+  }
+
+  const excluded = (extensions: Extension[], filter: string[]) =>
+    extensions.filter((extension) => !filter.includes(extension.localIdentifier))
+  if (extensionsExcluded && extensionsExcluded.length > 0) {
+    app.extensions.function = excluded(app.extensions.function, extensionsExcluded) as FunctionExtension[]
+    app.extensions.ui = excluded(app.extensions.function, extensionsExcluded) as UIExtension[]
```
[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/app/src/cli/services/dev.ts:467]: As you can see in the `How to test the changes` I have only executed manual tests with `ui` and `themes` extensions

**Comment by pepicrft** [comment]
@isaacroldan, @alvaro-shopify and I me to talk about whether this is a good approach. We concluded that the nature of `dev` is interactive, and that even if we could get it working as a service in Spin, the setup would end up being brittle, hard to reason about, and not very portable across Spin constellations.
We recommended the forms team to use the CLI interactively, which in other words means they'll have to run `yarn dev` after creating the Spin constellation and having all the services running. We might be able to revisit this in the future once all the extensions have moved to the extensions framework because that means we'll be able to authorize all the operations through the Partners' API.

**Pull request closed without merging**